Python/pystate.c

```diff
@@ -171,6 +171,65 @@ _PyRuntimeState_Fini(_PyRuntimeState *runtime)
 }
 
 #ifdef HAVE_FORK
+_Py_IDENTIFIER(stdout);
+_Py_IDENTIFIER(stderr);
+
+static int stdio_at_fork_reinit(_Py_Identifier *key)
+{
+
+    int ret = 0;
+
+    PyObject *stdio, *closed, *isatty, *buffer, *result;
+
+    stdio = _PySys_GetObjectId(key);
+
+    _Py_IDENTIFIER(closed);
+    if (_PyObject_LookupAttrId(stdio, &PyId_closed, &closed) < 0) {
+        PyErr_Clear();
+        goto end;
+    }
+
+    if (Py_IsTrue(closed)) {
+        goto end;
```

Comment:
Doesn't this leak `stdio` and `closed`?

Reply:
Re: leaking a reference to `sys.stdout` and `sys.stderr`.

Ok, so correct me if I'm wrong (I'm new to this), but when we say that something is leaking a reference, we mean the function we're calling causes a `Py_INCREF` to the `PyObject` it returns, which we then fail to `Py_DECREF` before returning (right?).

If so, then I don't think `_PySys_GetObjectId` causes an `INCREF`.

Unless I'm misunderstanding, this should print two different values if we're leaking a reference:

```
stdio = _PySys_GetObjectId(key);
fprintf(stderr, "stdout before: %ld\t", stdio->ob_refcnt);
stdio = _PySys_GetObjectId(key);
fprintf(stderr, "stdout after: %ld\n", stdio->ob_refcnt);
```

Same deal here:

```
isatty = _PyObject_CallMethodIdNoArgs(stdio, &PyId_isatty);
fprintf(stderr, "isatty before: %ld\t", isatty->ob_refcnt);
isatty = _PyObject_CallMethodIdNoArgs(stdio, &PyId_isatty);
fprintf(stderr, "isatty after: %ld\n", isatty->ob_refcnt);
```

When I add the above lines and run the tests, I see output like the following:

```
stdout before: 6	stdout after: 6
isatty before: 1489	isatty after: 1490
stdout before: 7	stdout after: 7
isatty before: 1490	isatty after: 1491
stdout before: 6	stdout after: 6
isatty before: 1491	isatty after: 1492
stdout before: 7	stdout after: 7
isatty before: 1492	isatty after: 1493
```

So (am I reasoning right here?) we need to `DECREF` (or `XDECREF`, if the `PyObject *` might be `NULL`) the return values of functions who print two different numbers here, but not the others.

So in this case, we'd `XDECREF` `isatty` (and the others like it) but not `stdio`.

Thanks again for the help. This is super fun. And the C-API docs are helpful, but there's nothing better than talking to an actual person who knows this stuff.

:)

Reply:
You're right -- `_PySys_GetObjectId` calls `_PyDict_GetItemIdWithError`. `PyDict_GetItem` and friends, including `_PyDict_GetItemIdWithError`, return a borrowed reference (i.e., they don't INCREF what they return, and reference still officially belongs to the sys dictionary).